io.asgardeo.tomcat.saml.agent/src/main/java/io/asgardeo/tomcat/saml/agent/SAML2SSOAgentFilter.java

```diff
@@ -86,76 +86,122 @@ public void doFilter(ServletRequest servletRequest, ServletResponse servletRespo
+                    handleException(request, response, ssoAgentConfig, e);
+                    return;
+                }
                 return;
             } else if (resolver.isSAML2SSOResponse()) {
 
-                samlSSOManager = new SAML2SSOManager(ssoAgentConfig);
                 try {
+                    samlSSOManager = new SAML2SSOManager(ssoAgentConfig);
                     samlSSOManager.processResponse(request, response);
+                    if (!"true".equals(request.getAttribute(SSOAgentConstants.SHOULD_GO_TO_WELCOME_PAGE))) {
+                        response.sendRedirect(ssoAgentConfig.getSAML2().getACSURL());
+                    }
                 } catch (SSOAgentException e) {
+                    if (e instanceof InvalidSessionException) {
+                        // Redirect to the index page when session is expired or user already logged out.
+                        LOGGER.log(Level.FINE, "Invalid Session!", e);
+                        response.sendRedirect(filterConfig.getServletContext().getContextPath());
+                        return;
+                    }
                     handleException(request, response, ssoAgentConfig, e);
                     return;
                 }
 
             } else if (resolver.isSAML2ArtifactResponse()) {
 
-                samlSSOManager = new SAML2SSOManager(ssoAgentConfig);
                 try {
+                    samlSSOManager = new SAML2SSOManager(ssoAgentConfig);
                     samlSSOManager.processArtifactResponse(request);
                 } catch (SSOAgentException e) {
                     handleException(request, response, ssoAgentConfig, e);
                     return;
                 }
             } else if (resolver.isSLOURL()) {
 
-                samlSSOManager = new SAML2SSOManager(ssoAgentConfig);
-                if (resolver.isHttpPostBinding()) {
-
-                    boolean isPassiveAuth = ssoAgentConfig.getSAML2().isPassiveAuthn();
-                    ssoAgentConfig.getSAML2().setPassiveAuthn(false);
-                    String htmlPayload = samlSSOManager.buildPostRequest(request, response, true);
-                    ssoAgentConfig.getSAML2().setPassiveAuthn(isPassiveAuth);
-                    SSOAgentUtils.sendPostResponse(request, response, htmlPayload);
-
-                } else {
-                    //if "SSOAgentConstants.HTTP_BINDING_PARAM" is not defined, default to redirect
-                    boolean isPassiveAuth = ssoAgentConfig.getSAML2().isPassiveAuthn();
-                    ssoAgentConfig.getSAML2().setPassiveAuthn(false);
-                    String redirectUrl = samlSSOManager.buildRedirectRequest(request, true);
-                    ssoAgentConfig.getSAML2().setPassiveAuthn(isPassiveAuth);
-                    response.sendRedirect(redirectUrl);
+                try {
+                    samlSSOManager = new SAML2SSOManager(ssoAgentConfig);
+                    if (resolver.isHttpPostBinding()) {
+
```

Comment:
let's remove the blank line

Reply:
Addressed with https://github.com/asgardeo/asgardeo-tomcat-saml-agent/pull/47/commits/dee9d739015733c812fbefca25ffab28bc6b2ffd